### DISCUSSION TITLE
Architectural Considerations for Product Clusters

### DISCUSSION THREAD

**Discussion by Ajay250496** [rook/rook#9871] [Q&A]
First off I want to say this project is awesome and have worked pretty flawlessly in our testing. For these reasons I want to move forward with deploying it on prod however I am now in the process of making some architectural decisions and want to make sure I've considered all factors before moving forward. I read through a few articles online, some documentation provided, and they all have been very helpful but I am still a bit confused on what would be the best way to deploy rook-ceph.

My cluster is a fairly simple cluster. We run a 4 node EKS cluster spanning a region with AWS which is runs a about 10 of the same application configured for 10 different clients. At current state we use PVC on EBS to drive our persistent storage. This achitecture obviously has flaws in storage as we have a zonal lock-in with EBS. For this reason we want to use rook-ceph to have multi-az pod failover functionality.

So far testing has been great and we can have all of this with the rook project. However I am trying to find out what the best way is to design this so that we are considering our current state and that it allows us to scale in the future without any issues.

My two burning questions are:
- What are best practices for the amount and size of OSDs? 
  - Do I simply calculate the amount of storage I currently have in my cluster, mutliply this by the amount of replication, and configure 1 OSD per node by dividing by 4? 
  - Should I create one OSD per application? (This causes management concerns) 
  - Am I overthinking this? I am considering how we scale as we add new applications as well. Lets say my overall storage consumption is about 300GB with a replication factor of 2, do I configure 600GB storage across 4 nodes?
- What sort of considerations do I have when scaling my cluster in the future. Lets say my requirement now jumps to 20 clients and I have to scale to 8 nodes. At current state its as simple as adding the nodes to the cluster. If i do this and Rook-ceph is running in this cluster will it scale automatically and I would just have to update the deployment to include more OSDs?

Apologies if these are answered elsewhere I have read through a few documentation and while they all provide great info I could not find the answers to these outright. Any thoughts are welcomed :)

PS: I am aware I may be drawing a direct correlation to application storage and the OSDs and ignoring the storage management layer here...so I will look further into this to make sure its not just a misunderstanding

**Comment by travisn** [comment]
The OSD topology is critical to consider. Rook will automatically pick up on the node labels in AWS to see which zones the OSDs belong to. See the [OSD topology labels](https://rook.github.io/docs/rook/v1.8/ceph-cluster-crd.html#osd-topology). With this assumption, then consider:
- The `failureDomain` should be set to `zone` so data is always spread across different zones, even if there are multiple OSDs in a single zone
- Use the cluster-on-pvc.yaml example to create the cluster, with OSDs backed by ebs volumes. Then OSD pods won't be tied to a specific node, they will just need to be on a node in the same zone as the ebs volume, which is done automatically. To keep control about how many OSDs per AZ, I'd recommend creating a `storageClassDeviceSet` per AZ, with `nodeAffinity` specified for the `preparePlacement` to that AZ.
- OSD sizes are commonly up to 4Ti, but there is no hard limit. You can also scale up the OSD sizes at a later time by updating the volumeClaimTemplate size in the cephcluster CR, and rook should increase the sizes automatically. Or you can increase the `storageClassDeviceSet.count`, and rook would add more OSDs

Hope this helps!

**Reply by BlaineEXE** [reply]
I would also suggest considering network throughput. With 600GB across 4 nodes, I'm not sure you need to think too hard about it, but for scaling up/out it becomes more of a consideration. There is an old but still relevant overview at the top of this document: https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/1.3/html/hardware_guide/networking_recommendations

If you have 300GB of data with replica 2 (600GB total), the total size of your ceph cluster should be larger than 600GB to accommodate the natural tendency for data to not balance perfectly across all OSD disks. Typically, ceph will complain about being "near full" when the used storage is 85% of available. I believe this is on a per-osd basis. 706+GB may be a better target to serve 300GB of storage ( 600GB / .85 = 706GB ), if I did that math right. You also can raise the near full (and full) ratios that Ceph uses internally, but OSDs do need some overhead.
https://stackoverflow.com/questions/58127536/why-ceph-turns-status-to-err-when-there-is-still-available-storage-space

One OSD per application isn't necessary, but you may want to consider whether you have throughput requirements for your applications. The network or the disks may be a bottleneck. For 4 nodes with 1gbit network interfaces, the throughput for each client to the ceph cluster with all clients going full-blast is something like 200mbit.
